src/devicecontroller/DefaultDeviceController.ts

```diff
@@ -552,6 +552,44 @@ export default class DefaultDeviceController implements DeviceControllerBasedMed
     );
   }
 
+  static getIntrinsicDeviceId(device: Device): string | null {
+    if (device === undefined) {
+      return undefined;
+    }
+
+    if (device === null) {
+      return null;
+    }
+
+    if (typeof device === 'string') {
+      return device;
+    }
+
+    if ((device as MediaStream).id) {
+      return (device as MediaStream).id;
+    }
+
+    const constraints: MediaTrackConstraints = device as MediaTrackConstraints;
+    if (constraints.deviceId === undefined) {
+      return undefined;
+    }
+
+    if (constraints.deviceId === null) {
+      return null;
+    }
+
+    if (typeof constraints.deviceId === 'string') {
+      return constraints.deviceId;
+    }
+
+    const deviceIdConstraint: ConstrainDOMStringParameters = constraints.deviceId as ConstrainDOMStringParameters;
+    if (typeof deviceIdConstraint.exact === 'string') {
+      return deviceIdConstraint.exact;
+    }
```

Comment:
Strictly speaking, `.deviceId` could be `instanceof Array`, and with this code it will fall through to the `undefined` case. You should probably handle that explicitly.

Reply:
Updated.